### DISCUSSION TITLE
What is the process of soft linking?

### DISCUSSION THREAD

**Discussion by classmatewu** [pnpm/pnpm#4365] [Q&A] [answered]
In this official article [Flat node_modules is not the only way](https://www.pnpm.cn/blog/2020/05/27/flat-node-modules-is-not-the-only-way), there is such a sentence:  
`The trick is that express is just a symlink. When Node.js resolves dependencies, it uses their real locations, so it does not preserve symlinks.`.   
So my question is why I can't see this soft link file, how is this link process defined, and how can I see the file of this soft link process?

**Comment by classmatewu** [comment]
How Soft Links Link to Hard Links?
@zkochan I'm very sorry to bother you here, I'm a pnpm lover, I didn't find the reason I wanted, I came here to ask you

**Comment by zkochan** [comment]
I don't understand what you don't see. Attach a screenshot of your node_modules folder.

**Comment by classmatewu** [comment, marked as answer]
![image](https://user-images.githubusercontent.com/44048381/155144937-650f5a27-c429-446b-b3de-540888b176b2.png)

**Reply by zkochan** [reply]
that folder is a symlink. Maybe your IDE doesn't visualize it as a symlink. But it is a symlink. The content of that folder is in `.pnpm/express@...`

**Reply by classmatewu** [reply]
Is it the express folder, or is express/node_modules a soft link folder

**Reply by zkochan** [reply]
express is the symlink. node_modules is a real folder but all files inside node_modules are symlinks

**Reply by classmatewu** [reply]
> Maybe your IDE doesn't visualize it as a symlink

how make my ide visualize it as a symlink. my ide is vscode. thank a lot.

**Reply by zkochan** [reply]
I don't know

this is how I see symlinks

![image](https://user-images.githubusercontent.com/1927579/155146521-086d77a7-1b9d-4544-9729-202bec150b8e.png)

**Reply by classmatewu** [reply]
> the `express` folder is a symlink

But why the `express/lib`、`express/readme.md`、`express/package.json` is real file

**Reply by zkochan** [reply]
they are hard links. Hard links look like real files but use the same disk space as files from the content-addressable store.

**Reply by classmatewu** [reply]
oh, can a folder is a symlink, but its sub file is hard link?

**Reply by zkochan** [reply]
yes

**Reply by classmatewu** [reply]
I'm a curious newbie so please bear with me

**Reply by classmatewu** [reply]
thank a lot. the pnpm is very amazing.
